Add a message for users running tf.js in node without the node backend

dsmilkov opened the pull request:
This PR adds a user-friendly message to remind people that we have node backend whenever they use tf.js in node with the vanilla cpu backend.

Also fixes a problem where a message saying `document is not defined` shows up at the top of the program. This was a side-effect of trying to access `document` when registering a WebGL backend.

<img width="719" alt="screen shot 2018-06-06 at 9 46 25 pm" src="https://user-images.githubusercontent.com/2294279/41073866-3fc504d4-69d3-11e8-98cf-17a08e44a3f1.png">


DOC

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-core/1085)
<!-- Reviewable:end -->

caisq commented:
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---

*[src/kernels/backend_cpu.ts, line 56 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN-KCWDnoUPYZNwiQM:-LEN-KCWDnoUPYZNwiQN:b-qqb1ro) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L56)):*
> ```
> 👋
> ```

What would this show up as in terminals that don't support unicode? Probably something like a question mark, as what I found when I tried it in non-unicode xterm. While I appreciate the lighthearted style, consider using something more robust.  :)

---

*[src/kernels/backend_cpu.ts, line 57 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN0HVU7t6x4BXHf2_w:-LEN0HVU7t6x4BXHf2_x:booc318) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L57)):*
> ```
> speed up 
> ```

Grammar nit: "speech up" is transitive. Maybe say "speed things up" or "speech computation up."

---

*[src/kernels/backend_cpu.ts, line 57 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN0XQKD1WU_VgkZvVX:-LEN0XQKD1WU_VgkZvVY:b-qn15hl) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L57)):*
> ```
>  backend
> ```

Grammar nit: add comma after "backend".

---

*[src/kernels/backend_cpu.ts, line 59 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN07KK3Yey1h3ygsIo:-LEN07KK3Yey1h3ygsIp:bx0gn9c) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L59)):*
> ```
> `npm i @tensorflow/tfjs-node`
> ```

Not only this, but also something like `import '@tensorflow/tfjs-node'` is required in the source code - is that right?

Also, should you say something about CUDA support?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/kernels/backend_cpu.ts, line 59 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN07KK3Yey1h3ygsIo:-LEN25-A0VTX82XeXaFJ:b-m7t02d) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L59)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> `npm i @tensorflow/tfjs-node`
> ```

Not only this, but also something like `import '@tensorflow/tfjs-node'` is required in the source code - is that right?

Also, should you say something about CUDA support?
</blockquote></details>

One idea, you can do a system call like `which nvidia-smi` to check whether nvidia-smi is on the path and if so, conditionally print some messages about CUDA tfjs-node.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Thanks for the review. PTAL.

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/kernels/backend_cpu.ts, line 56 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN-KCWDnoUPYZNwiQM:-LEN1FALE_90DMWbEtZl:btgwy84) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L56)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> 👋
> ```

What would this show up as in terminals that don't support unicode? Probably something like a question mark, as what I found when I tried it in non-unicode xterm. While I appreciate the lighthearted style, consider using something more robust.  :)
</blockquote></details>

It shows up as ö, which is not bad. Non-unicode terminals are getting more rare. I think we can afford some emoji in 2018.

---

*[src/kernels/backend_cpu.ts, line 57 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN0HVU7t6x4BXHf2_w:-LEN1ifc4uy9aWNmsYXK:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L57)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> speed up 
> ```

Grammar nit: "speech up" is transitive. Maybe say "speed things up" or "speech computation up."
</blockquote></details>

Done.

---

*[src/kernels/backend_cpu.ts, line 57 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN0XQKD1WU_VgkZvVX:-LEN1sTw-TwkgAz8vyHG:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L57)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
>  backend
> ```

Grammar nit: add comma after "backend".
</blockquote></details>

Done.

---

*[src/kernels/backend_cpu.ts, line 59 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN07KK3Yey1h3ygsIo:-LEN1uyt4XAuoYKs9hg6:bve7drr) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L59)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

One idea, you can do a system call like `which nvidia-smi` to check whether nvidia-smi is on the path and if so, conditionally print some messages about CUDA tfjs-node.
</blockquote></details>

Added CUDA. I did say "and import it in the program", but I added require('.....') to be more concise. Note I didn't use import '...' since that's es6 js. require() will work on all node versions. FYI, I didn't mention tf.setBackend('tensorflow') since that is going away in a upcoming PR (the import will do that for you).

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-:-LEN4onOEerNOQrOTaih:b-k49e1v)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/kernels/backend_cpu.ts, line 59 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN07KK3Yey1h3ygsIo:-LEN63kqA4CKhbYpPuWI:b-o2urf3) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L59)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Added CUDA. I did say "and import it in the program", but I added require('.....') to be more concise. Note I didn't use import '...' since that's es6 js. require() will work on all node versions. FYI, I didn't mention tf.setBackend('tensorflow') since that is going away in a upcoming PR (the import will do that for you).
</blockquote></details>

Leave `which nvidia-smi` for a future time. It will be tricky to conditionally import node utility libraries in tfjs-core that execute a system call while still making sure that the browserified bundle doesn't have those imports.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085)*
<!-- Sent from Reviewable.io -->

caisq commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

Thanks!

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-:-LENCwNKED6vgNWKVnMa:b-afugb9)*
<!-- Sent from Reviewable.io -->

nkreeger commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[src/kernels/backend_cpu.ts, line 59 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN07KK3Yey1h3ygsIo:-LENeT6v0sMfXkifH9bJ:b-f7bgcu) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L59)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Leave `which nvidia-smi` for a future time. It will be tricky to conditionally import node utility libraries in tfjs-core that execute a system call while still making sure that the browserified bundle doesn't have those imports.
</blockquote></details>

CUDA is Linux only and still a pain to install. Lots of Linux users don't have a CUDA-capable card - so let's not sweat that here. I think if users know about CUDA + `nvidia-smi` (which is linux only) they'll be able to find `@tensorflow/tfjs-node-gpu`.

I'd advise dropping this for now. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[src/kernels/backend_cpu.ts, line 59 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1085#-LEN07KK3Yey1h3ygsIo:-LEPV7hIC1oJK-SRsUWd:bf7sw1a) ([raw file](https://github.com/tensorflow/tfjs-core/blob/0f95d743c472c93b4c487cd7cafc4aa3f8aa3ec6/src/kernels/backend_cpu.ts#L59)):*
<details><summary><i>Previously, nkreeger (Nick Kreeger) wrote…</i></summary><blockquote>

CUDA is Linux only and still a pain to install. Lots of Linux users don't have a CUDA-capable card - so let's not sweat that here. I think if users know about CUDA + `nvidia-smi` (which is linux only) they'll be able to find `@tensorflow/tfjs-node-gpu`.

I'd advise dropping this for now. 
</blockquote></details>

I'll still keep the tiny plug for CUDA (in case someone has it, they should know we support it) but not going into details. People that don't know what CUDA is will either 1) ignore the "if you have CUDA" message, or 2) go google it and learn about it, which is also good

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1085)*
<!-- Sent from Reviewable.io -->
